feat: convert to TypeScript

aminya opened the pull request:
Closes #252 #253 #254

Includes the types from this PR. We can remove them here once that PR is merged.
https://github.com/steelbrain/exec/pull/103

aminya committed:
feat: install typescript + remove flow/babel

aminya committed:
chore: use eslint-config-atomic to support TypeScript

aminya committed:
feat: convert to TypeScript

```
pnpm i -g @khanacademy/flow-to-ts
flow-to-ts --write --delete-source .\src\*.js --prettier
```

aminya committed:
chore: fix return type of generateRange

aminya committed:
fix: use FS.constants instead of non-existing FS properties

aminya committed:
fix: use node's promisified fs function

aminya committed:
chore: add prettier

aminya committed:
fix: eslint fix

aminya committed:
chore: fix the input type of getWordRegexp

aminya committed:
fix: use FSP.access in fileExists

aminya committed:
chore: more strict tsconfig

aminya committed:
test: test the built files

aminya committed:
chore: use files entry instead of npmignore

aminya committed:
fix: update dependencies

aminya committed:
ci: fix CI errors by using GitHubActions

aminya committed:
Revert "ci: fix CI errors by using GitHubActions"

This reverts commit 409391f0be93301ecf3f89f3f7e19b941dbc317f.

aminya committed:
chore: fix the typeof NamedRegexp

aminya committed:
chore: fix the type of filePaths

aminya committed:
fix: lineStart, colStart, lineEnd, colEnd must be numbers

aminya committed:
chore: add types for tmp

aminya committed:
chore: add types for grim

aminya committed:
chore: types for sb-exec

aminya committed:
chore: fix error type in wrapExec

aminya committed:
fix: fix the sb-exec types and execWrap

aminya committed:
fix: catch the null bugs if lineText is undefined

aminya committed:
fix: help TypeScript infer colStart type

aminya committed:
fix: kill might not exist on spawned

aminya committed:
fix: check if name is array

aminya committed:
fix: fix the type of match inside parse

aminya commented:
@steelbrain Please take a look at this.

aminya committed:
chore: fix a couple of eslint warnings

steelbrain commented on .eslintrc.json:
```diff
@@ -1,6 +1,4 @@
-  "globals": {
-    "atom": true
-  }
+  "extends": "eslint-config-atomic",
```
Can we replace this with ESLint config Airbnb instead?

aminya replied:
I am reluctant to use the eslint configurations that ban certain syntax or patterns without adding any benefit to the project. As an example, Airbnb configuration, specifically, bans using `for of` in favor of `forEach`, although I have provided [numerous benchmarks](https://github.com/aminya/typescript-optimization) that disproves their claims about their Eslint rules.
In contrast, eslint-config-atomic is only made up of "constructive rules" that are backed up with benchmarks and research and respects the developer's choices for different programming paradigms.

P.S: the main reason I used [eslint-config-atomic](https://github.com/atom-community/eslint-config-atomic/tree/master/src) in this project was to get out-of-the-box support for all the languages such as TypeScript, which the previous config was failing to parse.

steelbrain reviewed (commented):
Thank you so much for working on this!

Appreciate the work here. I have some reservations about some of the tooling choices, can we please use npm or yarn instead of pnpm? Personally love pnpm but for public projects, it adds a barrier to entry IMO

Also some of the changes may be behavior changes, although they look harmless from a first glance, but it's possible they'll behave differently. Can we split the rewrite to TS and those changes into separate PRs or do you think they should go together?

aminya commented:
> Thank you so much for working on this!
> 

You're welcome!

> Appreciate the work here. I have some reservations about some of the tooling choices, can we please use npm or yarn instead of pnpm? Personally love pnpm but for public projects, it adds a barrier to entry IMO
> 

That's exactly why I removed the `package-lock.json` file and added the `npmrc` file. This gives people the freedom to use any package manager they like such as npm, yarn, or pnpm. I actually didn't intend to commit the `pnpm-lock.yaml`, but using the current config, `pnpm` handles out-of-date lock-file without issue, so people can update dependencies without breaking anything. I can remove this lock file if you prefer that.


> Also some of the changes may be behavior changes, although they look harmless from a first glance, but it's possible they'll behave differently. Can we split the rewrite to TS and those changes into separate PRs or do you think they should go together?


Yes, I always keep my commits specific, completely composable, and incremental to allow future rebasing, reverting, and cherry-picking. So, if you can give me the commit range that you would like me to move, I will happily move it to a follow-up PR. What I did in this PR are:
- Convert to TypeScript
- Fixing the tsc error
- Fixing eslint errors

I kept everything backward compatible, but still, feel free to give me the range that you would like me to move.

aminya committed:
chore: use prettier-config-atomic

aminya commented:
Any feedback on the answers? @steelbrain

steelbrain commented:
@aminya Generally looks good to me, I'll run/test this and then post further feedback (if any) then we can merge
